src/AbstractOperations/binary_operations.jl

```diff
@@ -187,6 +193,40 @@ macro binary(ops...)
     return expr
 end
 
+#####
+##### BinaryOperations with ZeroField
+#####
+
+import Base: +, -, *, /, ==
+using Oceananigans.Fields: ZeroField, ConstantField
+
+==(::ZeroField, ::ZeroField) = true
+
+==(zf::ZeroField, cf::ConstantField) = 0 == cf.constant
+==(cf::ConstantField, zf::ZeroField) = ==(zf, cf)
+==(c1::ConstantField, c2::ConstantField) = c1.constant == c2.constant
+
++(a::ZeroField, b::AbstractField) = b
++(a::AbstractField, b::ZeroField) = +(b, a)
++(a::ZeroField, b::Number) = ConstantField(b)
++(a::Number, b::ZeroField) = +(b, a)
+
+-(a::ZeroField, b::AbstractField) = -b
+-(b::AbstractField, a::ZeroField) = b
+-(a::ZeroField, b::Number) = ConstantField(-b)
+-(b::Number, a::ZeroField) = ConstantField(b)
+
+*(a::ZeroField, b::AbstractField) = a
+*(b::AbstractField, a::ZeroField) = *(a, b)
+*(a::ZeroField, b::Number) = a
+*(b::Number, a::ZeroField) = *(a, b)
+
+/(a::ZeroField, b::AbstractField) = a
+/(b::AbstractField, a::ZeroField) = ConstantField(Inf)
```

Comment:
```suggestion
/(b::AbstractField, a::ZeroField) = convert(eltype(b), Inf)
```

Reply:
Note that `Inf` is type `Float64`.

Reply:
So we don't want to use `Inf` because this can cause unintended type promotion for `Float32` grids, etc.

Reply:
Good point. However, with that code a user can't use integers since there's no integer Inf:

```julia
julia> 2/ZeroField()
ERROR: InexactError: Int64(Inf)
Stacktrace:
 [1] Int64
   @ ./float.jl:788 [inlined]
 [2] convert(#unused#::Type{Int64}, x::Float64)
   @ Base ./number.jl:7
 [3] /(a::Int64, b::ZeroField{Int64, 3})
   @ Oceananigans.AbstractOperations ~/repos/Oceananigans.jl2/src/AbstractOperations/binary_operations.jl:228
 [4] top-level scope
   @ REPL[19]:1
 [5] top-level scope
   @ ~/.julia/packages/CUDA/DfvRa/src/initialization.jl:52

julia> convert(eltype(2), Inf)
ERROR: InexactError: Int64(Inf)
Stacktrace:
 [1] Int64
   @ ./float.jl:788 [inlined]
 [2] convert(#unused#::Type{Int64}, x::Float64)
   @ Base ./number.jl:7
 [3] top-level scope
   @ REPL[20]:1
 [4] top-level scope
   @ ~/.julia/packages/CUDA/DfvRa/src/initialization.jl:52
```

Are we okay with that?

Reply:
I'm not sure what else would be correct

Reply:
I guess an alternative way would be to define functions like

```julia
/(a::Number, b::ZeroField) = ConstantField(a/convert(eltype(a), 0))
```

and let Julia handle it. That way the output would follow the usual Julia conventions which rely on `propagate`:

```julia
julia> 2/0.0
Inf
```

Reply:
After thinking about it, I do think the above way is preferred for `::Number/::ZeroField`. So I made that change and changed the `AbstractField` division according your suggestion.

If you're okay with this, I think this is ready for merging.

Reply:
Looks good